src/controller/c/robot.c

```diff
@@ -1130,28 +1130,51 @@ static char *compute_socket_filename() {
     free(robot_name);
     return socket_filename;
   }
+
+#ifndef _WIN32
+  const char *username = wbu_system_getenv("USER");
+  if (username == NULL || username[0] == '\0') {
+    username = wbu_system_getenv("USERNAME");
+    if (username == NULL || username[0] == '\0') {
+      fprintf(stderr, "Error: USER or USERNAME environment variable not set, falling back to 'default' username.");
+      username = "default";
+    }
+  }
+#endif
+
   // extern controller case
   // parse WEBOTS_CONTROLLER_URL to extract protocol, host, port and robot name
   const char *TMP_DIR = wbu_system_tmpdir();
   char *WEBOTS_CONTROLLER_URL = (char *)wbu_system_getenv("WEBOTS_CONTROLLER_URL");
+  // either the WEBOTS_CONTROLLER_URL is not defined, empty or contains only a robot name
+  // default to the most recent /tmp/webots/username/* folder (/tmp/webots-* on Windows)
   if (WEBOTS_CONTROLLER_URL == NULL || WEBOTS_CONTROLLER_URL[0] == 0 || strstr(WEBOTS_CONTROLLER_URL, "://") == NULL) {
-    // either the WEBOTS_CONTROLLER_URL is not defined, empty or contains only a robot name
-    // default to the most recent /tmp/webots-* folder
-    const int TMP_DIR_LENGTH = strlen(TMP_DIR);
-    DIR *dr = opendir(TMP_DIR);
+#ifndef _WIN32
+    const int WEBOTS_TMP_DIR_length = strlen(TMP_DIR) + strlen(username) + 9;  // TMP_DIR + '/webots/' + username
+    char *WEBOTS_TMP_DIR = malloc(WEBOTS_TMP_DIR_length);
+    snprintf(WEBOTS_TMP_DIR, WEBOTS_TMP_DIR_length, "%s/webots/%s", TMP_DIR, username);
+#else
+    const int WEBOTS_TMP_DIR_length = strlen(TMP_DIR);
+    char *WEBOTS_TMP_DIR = strdup(TMP_DIR);
+#endif
+    DIR *dr = opendir(WEBOTS_TMP_DIR);
     if (dr == NULL) {
-      fprintf(stderr, "Error: cannot open directory %s\n", TMP_DIR);
+      fprintf(stderr, "Error: cannot open directory %s\n", WEBOTS_TMP_DIR);
       exit(EXIT_FAILURE);
     }
     struct stat filestat;
     double timestamp = 0.0;
     int number = -1;
     struct dirent *de;
     while ((de = readdir(dr)) != NULL) {
```

Comment:
We should probably `free(WEBOTS_TMP_DIR);` after this loop.

Reply:
It's weird that cppcheck didn't detect the missing free()...